src/reference/mod.rs

```diff
@@ -0,0 +1,367 @@
+//! that the stanza matched against.
+//!
+//! # Graph nodes
+//!
+//! Each node in the output graph is identified by a syntax node and a list of **_tag names_**.
+//! Graph nodes are automatically created when they are referenced.  Graph node expressions have
+//! one of the following formats:
+//!
+//!   - `[syntax node].[tag name]`
+//!   - `[graph node].[tag name]`
+//!
+//! For instance, all of the following are valid graph node expressions in our example stanza:
+//!
+//!   - `@id.node`
+//!   - `@id.node.another`
+//!
+//! For both expressions, the graph node's syntax node would be whichever node was captured by
+//! `@id` in the stanza's query.  They would have `node` and `node.another`, respectively, as their
+//! tag names.
+//!
+//! There can be at most one graph node for any combination of syntax node and tag names.  (The
+//! name of the capture used to identify the syntax node does not matter, what counts is which
+//! syntax node the capture refers to.)  If multiple stanzas refer to graph nodes with the same
+//! syntax node and tag names, those are "collapsed" into a single graph node.
+//!
+//! # Edges
+//!
+//! Edges are created via an `edge` statement, which specifies the two graph nodes that should be
+//! connected.  Edges are directed, and the `->` arrow in the `edge` statement indicates the
+//! direction of the edge.
+//!
+//! ``` tsg
+//! (import_statement name: (_) @name)
+//! {
+//!   edge @name.definition -> @name.reference
+//! }
+//! ```
+//!
+//! There can be at most one edge connecting any particular source and sink node in the graph.  If
+//! multiple stanzas create edges between the same nodes, those are "collapsed" into a single edge.
+//!
+//! # Attributes
+//!
+//! Graph nodes and edges have an associated set of **_attributes_**.  Each attribute has a string
+//! name, and a value.
+//!
+//! You add attributes to a node using an `attr` statement, which takes in a graph node expression,
+//! a string expression for the name of the attribute, and an expression of any type for the value
+//! of the attribute:
+//!
+//! ``` tsg
+//! (import_statement name: (_) @name)
+//! {
+//!   attr @name.definition "kind" = "module"
+//! }
+//! ```
+//!
+//! You add attributes to an edge by using the optional attributes clause of an `edge` statement:
+//!
+//! ``` tsg
+//! (import_statement name: (_) @name)
+//! {
+//!   edge @name.definition -> @name.reference, "precedence" = 10
+//! }
+//! ```
+//!
+//! Multiple stanzas can add attributes to the same graph node or edge.  Any attributes defined in
+//! those stanzas are **_combined_** into a single set of attributes by taking the union of their
+//! contents.  If both stanzas define an attribute with a particular name, then the corresponding
+//! values must either be:
+//!
+//!   - both sets, in which case the values are unioned together
+//!   - anything else, in which case the values must be equal
+//!
```

Comment:
What's the syntax like for a set in this DSL? I wonder if this would be helpful to add as an example in the `# Expressions` section above (along with an example showing an ordered list of values)?

Reply:
Added a description of this and lists